File: src/com/pilot51/lander/SeekBarPreference.java
package com.pilot51.lander;

import java.text.DecimalFormat;

import android.content.Context;
import android.preference.Preference;
import android.util.AttributeSet;
import android.view.View;
import android.widget.SeekBar;
import android.widget.SeekBar.OnSeekBarChangeListener;
import android.widget.TextView;

public class SeekBarPreference extends Preference implements OnSeekBarChangeListener {
	private SeekBar seekbar;
	private TextView valueText;
	private int steps, barMin, barMax;
	private float value, fDefault, min, max, increment;
	/** Increment as a divisor of 1 */
	private float subIncrement;
	/** Divide by this to convert actual value to visible bar value, or multiply for the other direction.<br />
	 * Useful, for example, to display a color value as a percentage when the actual range is 0-255. */
	private float scale;
	private String suffix;
	private DecimalFormat df = new DecimalFormat("0.##");

	public SeekBarPreference(Context context, AttributeSet attrs) {
		super(context, attrs);
		setLayoutResource(R.layout.preference_seekbar);
		String xmlns = "http://schemas.android.com/apk/res/com.pilot51.lander";
		max = attrs.getAttributeFloatValue(xmlns, "max", 0);
		min = attrs.getAttributeFloatValue(xmlns, "min", 0);
		fDefault = attrs.getAttributeFloatValue(xmlns, "defaultValue", 0);
		increment = attrs.getAttributeFloatValue(xmlns, "increment", 1);
		subIncrement = attrs.getAttributeFloatValue(xmlns, "subIncrement", 1);
		steps = attrs.getAttributeIntValue(xmlns, "steps", 0);
		scale = attrs.getAttributeFloatValue(xmlns, "valueScale", 1);
		suffix = attrs.getAttributeValue(xmlns, "suffix");
		if (suffix == null)
			suffix = "";
		if (steps > 1)
			barMax = steps;
		else barMax = convertToBarValue(max);
		barMin = convertToBarValue(min);
		setPersistent(true);
		setDefaultValue(fDefault);
	}

	@Override
	protected void onSetInitialValue(boolean restorePersistedValue, Object defaultValue) {
		value = getPersistedFloat(fDefault);
		if (!restorePersistedValue) persistFloat(value);
	}

	@Override
	protected void onBindView(View view) {
		super.onBindView(view);
		seekbar = (SeekBar) view.findViewById(R.id.seekbar);
		seekbar.setMax(barMax);
		seekbar.setOnSeekBarChangeListener(this);
		valueText = (TextView) view.findViewById(R.id.value);
		valueText.setText(df.format(value / scale) + suffix);
		seekbar.setProgress(convertToBarValue(value));
	}

	@Override
	public void onProgressChanged(SeekBar seekBar, int progress, boolean fromTouch) {
		value = convertToValue(progress >= barMin ? progress : barMin);
		valueText.setText(df.format(value / scale) + suffix);
	}

	@Override
	public void onStartTrackingTouch(SeekBar seekBar) {}

	@Override
	public void onStopTrackingTouch(SeekBar seekBar) {
		persistFloat(value);
	}
	
	protected void setToDefault() {
		this.value = fDefault;
		valueText.setText(df.format(value / scale) + suffix);
		seekbar.setProgress(convertToBarValue(value));
		persistFloat(value);
	}

	private float convertToValue(int value) {
		if (subIncrement > 1) return (min + (value / subIncrement)) * scale;
		else return (min + (value * increment)) * scale;
	}

	private int convertToBarValue(float value) {
		if (subIncrement > 1) return (int)((value - min) * subIncrement / scale);
		else return (int)((value - min) / increment / scale);
	}
}

File: android-app/src/main/java/com/pilot51/lander/SeekBarPreference.kt
package com.pilot51.lander

import android.content.Context
import android.util.AttributeSet
import android.widget.SeekBar
import android.widget.SeekBar.OnSeekBarChangeListener
import android.widget.TextView
import androidx.preference.Preference
import androidx.preference.PreferenceViewHolder
import java.text.DecimalFormat


class SeekBarPreference(
	context: Context, attrs: AttributeSet
) : Preference(context, attrs), OnSeekBarChangeListener {
	private lateinit var seekbar: SeekBar
	private lateinit var valueText: TextView
	private val sAttrs = context.obtainStyledAttributes(attrs, R.styleable.SeekBarPreference, 0, 0)
	private val steps = sAttrs.getInt(R.styleable.SeekBarPreference_seekSteps, 0)
	private val fDefault = sAttrs.getFloat(R.styleable.SeekBarPreference_seekDefaultValue, 0f)
	private val min = sAttrs.getFloat(R.styleable.SeekBarPreference_seekMin, 0f)
	private val max = sAttrs.getFloat(R.styleable.SeekBarPreference_seekMax, 0f)
	private val increment = sAttrs.getFloat(R.styleable.SeekBarPreference_seekIncrement, 1f)
	/** Increment as a divisor of 1 */
	private val subIncrement = sAttrs.getFloat(R.styleable.SeekBarPreference_seekSubIncrement, 1f)
	/**
	 * Divide by this to convert actual value to visible bar value, or multiply for the other direction.
	 * Useful, for example, to display a color value as a percentage when the actual range is 0-255.
	 */
	private val scale = sAttrs.getFloat(R.styleable.SeekBarPreference_seekValueScale, 1f)
	private val suffix = sAttrs.getString(R.styleable.SeekBarPreference_seekSuffix) ?: ""
	private val barMin = convertToBarValue(min)
	private val barMax = if (steps > 1) steps else convertToBarValue(max)
	private val df = DecimalFormat("0.##")
	private var value = 0f

	init {
		layoutResource = R.layout.preference_seekbar
		isPersistent = true
		setDefaultValue(fDefault)
	}

	@Deprecated("Deprecated in Java")
	override fun onSetInitialValue(restorePersistedValue: Boolean, defaultValue: Any?) {
		value = getPersistedFloat(fDefault)
		if (!restorePersistedValue) persistFloat(value)
	}

	override fun onBindViewHolder(holder: PreferenceViewHolder) {
		super.onBindViewHolder(holder)
		seekbar = holder.findViewById(R.id.seekbar) as SeekBar
		seekbar.max = barMax
		seekbar.setOnSeekBarChangeListener(this)
		valueText = holder.findViewById(R.id.value) as TextView
		valueText.text = df.format(value / scale.toDouble()) + suffix
		seekbar.progress = convertToBarValue(value)
	}

	override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromTouch: Boolean) {
		value = convertToValue(if (progress >= barMin) progress else barMin)
		valueText.text = df.format(value / scale.toDouble()) + suffix
	}

	override fun onStartTrackingTouch(seekBar: SeekBar) {}

	override fun onStopTrackingTouch(seekBar: SeekBar) {
		persistFloat(value)
	}

	override fun onClick() {
		seekbar.progress = if (seekbar.progress < seekbar.max) {
			seekbar.progress + 1
		} else barMin
		value = convertToValue(seekbar.progress)
		valueText.text = df.format(value / scale.toDouble()) + suffix
		persistFloat(value)
	}

	fun setToDefault() {
		value = fDefault
		valueText.text = df.format(value / scale.toDouble()) + suffix
		seekbar.progress = convertToBarValue(value)
		persistFloat(value)
	}

	private fun convertToValue(value: Int): Float {
		return if (subIncrement > 1) {
			(min + value / subIncrement) * scale
		} else (min + value * increment) * scale
	}

	private fun convertToBarValue(value: Float): Int {
		return if (subIncrement > 1) {
			((value - min) * subIncrement / scale).toInt()
		} else ((value - min) / increment / scale).toInt()
	}
}
